### PR TITLE
Faster reading cache

### PR DESCRIPTION
**Pull request by jackvreeken**
- [x] Merge #136  first. This goes on top.

### CODE REVIEW

File: src/pymoca/backends/casadi/api.py [hunk 1]
```diff
@@ -308,19 +346,29 @@ def load_model(model_folder: str, model_name: str, compiler_options: Dict[str, s
...
+                    actual_deps = {all_symbols[j] for j in duration_dependencies[i]}
+                    false_deps = deps - actual_deps
+
+                    if false_deps:
                         [dur] = ca.substitute(
                             [dur],
-                            [false_deps],
-                            [ca.DM(np.full(false_deps.size1(), np.nan))])
+                            list(false_deps),
+                            [ca.repmat(np.nan, *d.size()) for d in false_deps])
```
Comment (vdwees) [inline src/pymoca/backends/casadi/api.py]: Not 100% sure, but I think that the numpy call is faster, and that you can use it without casting to DM. But maybe there is a good reason to use repmat?

File: src/pymoca/backends/casadi/api.py [hunk 2]
```diff
@@ -191,6 +191,45 @@ def save_model(model_folder: str, model_name: str, model: Model,
...
+        if model.delay_states:
+
+            all_symbols = [model.time,
+                           *model._symbols(model.states),
+                           *model._symbols(model.der_states),
+                           *model._symbols(model.alg_states),
+                           *model._symbols(model.inputs),
+                           *model._symbols(model.constants),
+                           *model._symbols(model.parameters)]
+            all_hashes = [hash(x) for x in all_symbols]
```
Comment (vdwees) [inline src/pymoca/backends/casadi/api.py]: it is not immediately obvious why you are using a hash. I imagine it is faster to do a linear search on? Why not use an an ordered dict or something else with fast lookups?

File: src/pymoca/backends/casadi/api.py (hunk 1 above)
Comment (jackvreeken) [reply src/pymoca/backends/casadi/api.py]: image

repmat is faster than what used to be there. Numpy arrays are 10x faster, but I think CasADi will convert to DM internally. And either way it's likely that `substitute` will be the bottleneck.

Comment (vdwees) [reply src/pymoca/backends/casadi/api.py]: agreed.

File: src/pymoca/backends/casadi/api.py (hunk 2 above)
Comment (vdwees) [reply src/pymoca/backends/casadi/api.py]: Something like
```python
{x: i for i, x in enumerate(all_symbols)}
```

Comment (jackvreeken) [reply src/pymoca/backends/casadi/api.py]: Fixed.

**Pull request merged**